span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java

```
@@ -127,23 +151,88 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
       containsPIIFields.set(false);
 
-      tagKeys.stream()
-          .filter(tagKey -> tagsToRedact.contains(tagKey.toUpperCase()))
-          .peek(tagKey -> containsPIIFields.set(true))
-          .forEach(
-              tagKey ->
-                  attributeMap.put(
-                      tagKey,
-                      AttributeValue.newBuilder()
-                          .setValue(SpanNormalizerConstants.PII_FIELD_REDACTED_VAL)
-                          .build()));
+      try {
+        tagKeys.stream()
+            .filter(
+                tagKey ->
+                    tagKeysToRedact.contains(tagKey.toUpperCase())
+                        || attributeMap
+                            .get(tagKey)
+                            .getValue()
+                            .equals(SpanNormalizerConstants.PII_FIELD_REDACTED_VAL))
+            .peek(tagKey -> containsPIIFields.set(true))
+            .forEach(
+                tagKey -> {
+                  spanAttributesRedactedCounters
+                      .computeIfAbsent(
+                          PIIMatchType.KEY.toString(),
+                          k ->
+                              registerCounter(
+                                  SPAN_REDACTED_ATTRIBUTES_COUNTER,
+                                  Map.of("matchType", PIIMatchType.KEY.toString())))
+                      .increment();
+                  logSpanRedaction(tagKey, PIIMatchType.KEY);
+                  if (redactedAttributeValue == null) {
+                    redactedAttributeValue =
+                        AttributeValue.newBuilder()
+                            .setValue(SpanNormalizerConstants.PII_FIELD_REDACTED_VAL)
+                            .build();
+                  }
+                  attributeMap.put(tagKey, redactedAttributeValue);
+                });
+      } catch (Exception e) {
+        LOG.error("An exception occurred while sanitising spans with key match: ", e);
```

Lets re-throw this exception.

---

Discussed on slack